ruby/src/api.cc

```diff
@@ -0,0 +1,509 @@
+        LOAD_SYMBOL(ruby_init),
+        LOAD_SYMBOL(ruby_options),
+        LOAD_SYMBOL(ruby_cleanup),
+        _library(move(library))
+    {
+    }
+
+    api::~api()
+    {
+        // API is shutting down; free all remaining data objects
+        // Destructors may unregister the data object, so increment the iterator before freeing
+        for (auto it = _data_objects.begin(); it != _data_objects.end();) {
+            auto data = reinterpret_cast<RData*>(*it);
+            ++it;
+            if (data->dfree) {
+                data->dfree(data->data);
+                data->dfree = nullptr;
+            }
+        }
+        if (_initialized && _library.first_load() && cleanup) {
+            ruby_cleanup(0);
+        }
+    }
+
+    api* api::instance()
+    {
+        static unique_ptr<api> instance = create();
+        return instance.get();
+    }
+
+    unique_ptr<api> api::create()
+    {
+        lth_lib::dynamic_library library = find_library();
+        if (!library.loaded()) {
+            LOG_WARNING("could not locate a ruby library");
+            return nullptr;
+        } else if (library.first_load()) {
+            LOG_INFO("ruby loaded from \"%1%\".", library.name());
+        } else {
+            LOG_INFO("ruby was already loaded.");
+        }
+        try {
+            return unique_ptr<api>(new api(move(library)));
+        } catch (lth_lib::missing_import_exception& ex) {
+            LOG_WARNING(ex.what());
+            return nullptr;
+        }
+    }
+
+    void api::initialize()
+    {
+        if (_initialized) {
+            return;
+        }
+
+        // Prefer ruby_setup over ruby_init if present (2.0+)
+        // If ruby is already initialized, this is a no-op
+        if (ruby_setup) {
+            ruby_setup();
+        } else {
+            ruby_init();
+        }
+
+        LOG_INFO("using ruby version %1%", to_string(rb_const_get(*rb_cObject, rb_intern("RUBY_VERSION"))));
+
+        if (_library.first_load()) {
+            // Run an empty script evaluation
+            // ruby_options is a required call as it sets up some important stuff (unfortunately)
+            char const* opts[] = {
+                "ruby",
+                "-e",
+                ""
+            };
+
+            // Check for bundler; this is the only ruby option we support
+            string ruby_opt;
+            if (environment::get("RUBYOPT", ruby_opt) && boost::starts_with(ruby_opt, "-rbundler/setup")) {
+                environment::set("RUBYOPT", "-rbundler/setup");
+            } else {
+                // Clear RUBYOPT so that only our options are used.
+                environment::set("RUBYOPT", "");
+            }
+
+            ruby_options(sizeof(opts) / sizeof(opts[0]), const_cast<char**>(opts));
+        }
+
+        // Get the values for nil, true, and false
+        // We do this because these are not constant across ruby versions
+        _nil = rb_ivar_get(*rb_cObject, rb_intern("@facter_nil"));
```

Comment:
That may be calling a value it assumes to be nil, in which case the name facter_nil could be changed to leatherman_nil.